lib/Dialect/SV/Transforms/HWDeclareTypedecls.cpp

```diff
@@ -0,0 +1,87 @@
+//===- HWDeclareTypedecls.cpp - HW DeclareTypedecls Pass ------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "SVPassDetail.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWTypes.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SymbolTable.h"
+
+using namespace circt;
+using namespace hw;
+
+namespace {
+struct HWDeclareTypedeclsPass
+    : public sv::HWDeclareTypedeclsBase<HWDeclareTypedeclsPass> {
+  void runOnOperation() override;
+
+private:
+  void visitOperation(Operation *op);
+  void visitType(Operation *op, Type type);
+  ModuleOp parentModule;
+  OpBuilder *builder;
+};
+} // namespace
+
+void HWDeclareTypedeclsPass::visitOperation(Operation *op) {
+  for (auto type : op->getOperandTypes())
+    visitType(op, type);
+  for (auto type : op->getResultTypes())
+    visitType(op, type);
+  for (auto &region : op->getRegions())
+    for (auto &block : region.getBlocks())
+      for (auto type : block.getArgumentTypes())
+        visitType(op, type);
+}
+
+void HWDeclareTypedeclsPass::visitType(Operation *op, Type type) {
+  // Look for any TypeAliasTypes.
+  auto alias = type.dyn_cast<TypeAliasType>();
+  if (!alias)
+    return;
+
+  // If no scope exists, create one in the parent module.
+  StringRef scopeName = alias.getRef().getRootReference();
```

Comment:
Does the `typedecl` op verifier check that the symbol ref is only two deep? If not, it probably should.